### PR TITLE
Concurrent and independent use of Z Probe endstop and Z Min endstop.

### PR DESCRIPTION
**Pull request by croadfeldt**
Add ability to use Z Probe and Z Min endstops independently. This is done by using a custom pin for the Z Probe endstop while leaving the Z Min endstop as defined in the appropriate pins_XXXX.h files.
Added code to use Z probe only on bed calibration and leveling routines, unless Z_SAFE_HOMING is defined, then it's used for Z axis homing as well. Neither negates the use of Z min, which remains active at all times.
Added Z_PROBE_PIN to pins_RAMPS_13.h.
Added Z_PROBE_AND_ENDSTOP and Z_PROBE_ENDSTOP_XXXXXX items to match existing endstop defines to configuration.h.
Usage listed in Auto Bed Leveling section of configuration.h

### CODE REVIEW

**Comment by croadfeldt** [comment]
Sorry for leaving out a title in the Pull Request. Can that be changed? 

EDIT: Got it, updated the title to something useful.

File: Marlin/language.h [hunk 1]
```diff
@@ -132,6 +132,7 @@
 #define MSG_Z_MIN                           "z_min: "
 #define MSG_Z_MAX                           "z_max: "
 #define MSG_Z2_MAX                          "z2_max: "
+#define MSG_Z_PROBE							"z_probe: "
```
Comment (thinkyhead) [inline Marlin/language.h:141]: spaces, not tabs

File: Marlin/pins.h [hunk 2]
```diff
@@ -178,6 +178,10 @@
   #define Z_MIN_PIN          -1
 #endif
 
+#ifdef DISABLE_Z_PROBE_ENDSTOP
+  #define Z_PROBE_PIN        -1
```
Comment (thinkyhead) [inline Marlin/pins.h:191]: `#undef Z_PROBE_PIN` is needed here to squash a compiler warning.

**Comment by croadfeldt** [comment]
Don't merge this yet... Something broke when I merged with the main Development branch again. Did a G1 Z10 and it moved the wrong direction and hit the Z Min endstop. Then I did a G28 to home and it pushed it further, ignoring the endstop triggered. Not sure where the issue is, but to be safe, hold off until I can take a look further.

**Comment by thinkyhead** [comment]
This is actually the "fun" part…

**Comment by croadfeldt** [comment]
For some reason I disabled Z_RAISE_BEFORE_HOMING in Marlin_main.cpp if Z_PROBE_AND_ENDSTOP is enabled. This very well could explain at least part of the issue I had last night. I can't think of why we would ever not move up before do a Z homing, especially if we have a probe to deploy. Commit incoming for reverting that...

**Comment by croadfeldt** [comment]
There's still a bug somewhere. M119 is returning open on a NC trigger when nothing is connected to the signal pin. Hold off on the merge still... Trying to track it down.

**Comment by croadfeldt** [comment]
Found the issue with the Z_PROBE not triggering and fixed it. G29 is still broken in that it doesn't raise the probe except at the start. But I don't believe that issue is in here.

There is a question though about the use of the Z Probe for homing. Is there any reason at all not to use it for homing at all times?

I can't think of a reason, this relegates the Z_MIN or Z_MAX endstop to a safety net basically. In which case there is a good argument for just separating the Z_Probe into it's own fully separate space. It's %80 of the way there now, but the name should change and the documentation should be updated to fully differentiate it from Z_MIN or Z_MAX.

I've documented in the code where the final pieces to fully separate it are, or at least the ones I found looking for this bug.

As it is, one more commit coming for this bug squash and some code comments.

There is another issue that I will bring up outside of here though. It could be a config issue, but it appears to be a break in behavior between release 1.0.2 and this branch.

**Comment by croadfeldt** [comment]
@thinkyhead The more I work through this, the more it becomes clear to just make Z Probe Endstop unique and have the ability to set a custom pin for it. Then use Z_PROBE_PIN instead of Z_MIN_PIN for all homing and calibration if it's enabled. So that's what I'm going to do.

**Comment by croadfeldt** [comment]
All set. No more commits coming unless Z_PROBE_PIN should be added to the rest of the PINS_XXXX.h files or bugs are found. Merge away if it's useful.

**Comment by thinkyhead** [comment]
@croadfeldt The build is failing, and it has conflicts so it can't be merged. You will need to fetch the latest upstream changes from Development and merge them with your changes. You should first create a new branch for this feature, instead of using Development, because it will save many headaches.

**Comment by croadfeldt** [comment]
Closing this PR in preference for a nice shiny lean new one.

**Comment by gralco** [comment]
> There is a question though about the use of the Z Probe for homing. Is there any reason at all not to use it for homing at all times?
> 
> I can't think of a reason, this relegates the Z_MIN or Z_MAX endstop to a safety net basically. In which case there is a good argument for just separating the Z_Probe into it's own fully separate space. It's %80 of the way there now, but the name should change and the documentation should be updated to fully differentiate it from Z_MIN or Z_MAX.

Those who use a capacitive, inductive, or optical switch for their Z_PROBE_PIN along with a dedicated Z_MIN_PIN sensor would not want their probe to trigger a Z home signal. These types of switches would normally be used at a Z level that is greater than the nozzle's (around Z=0 at home).

**Pull request closed without merging**